src/keygen/output.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2022-2023 Bolt Labs Holdings, Inc
+//
+// This source code is licensed under both the MIT license found in the
+// LICENSE-MIT file in the root directory of this source tree and the Apache
+// License, Version 2.0 found in the LICENSE-APACHE file in the root directory
+// of this source tree.
+
+use std::collections::HashSet;
+
+use crate::{
+    errors::{CallerError, InternalError, Result},
+    keygen::keyshare::{KeySharePrivate, KeySharePublic},
+    utils::CurvePoint,
+};
+
+use k256::ecdsa::VerifyingKey;
+use tracing::error;
+
+/// Output type from key generation, including all parties' public key shares,
+/// this party's private key share, and a bit of global randomness.
+#[derive(Debug, Clone)]
+pub struct Output {
+    public_key_shares: Vec<KeySharePublic>,
+    private_key_share: KeySharePrivate,
+    rid: [u8; 32],
+}
+
+impl Output {
+    /// Construct the generated public key.
+    pub fn public_key(&self) -> Result<VerifyingKey> {
+        // Add up all the key shares
+        let public_key_point = self
+            .public_key_shares
+            .iter()
+            .fold(CurvePoint::IDENTITY, |sum, share| sum + *share.as_ref());
+
+        VerifyingKey::from_encoded_point(&public_key_point.0.to_affine().into()).map_err(|_| {
+            error!("Keygen output does not produce a valid public key.");
+            InternalError::InternalInvariantFailed
+        })
+    }
+
+    pub(crate) fn public_key_shares(&self) -> &[KeySharePublic] {
+        &self.public_key_shares
+    }
+
+    pub(crate) fn private_key_share(&self) -> &KeySharePrivate {
+        &self.private_key_share
+    }
+
+    /// This could be made public if appropriate
+    #[cfg(test)]
+    pub(crate) fn rid(&self) -> &[u8; 32] {
+        &self.rid
+    }
+
+    /// Create a new `Output` from its constitutent parts.
+    ///
+    /// This method should only be used with components that were previously
+    /// derived via the [`Output::into_parts()`] method; the calling application
+    /// should not try to form public and private key shares independently.
+    ///
+    /// The provided components must satisfy the following properties:
+    /// - There is a valid key pair -- that is, the public key corresponding to
+    ///   the private key share must be contained in the list of public shares.
```

Comment:
We use unique sets around a few places and have to check for their uniqueness. Should we consider adding a `UniqueSet`/`UniqueSet<T>` which does the validation once and is guaranteed to be unique? This could cut down on documenting this requirement and checking it in all public APIs

Reply:
Hm yeah, that's a nice idea. I made #483